iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -387,22 +415,59 @@ static void legalizeDispatchWorkgroupOperands(
       // TODO(nicolasvasilache): do something useful.
       continue;
     }
-    operand.replaceAllUsesExcept(
-        repl, SmallPtrSet<Operation *, 8>(uses.begin(), uses.end()));
+    map.map(operand, repl);
   }
 
-  // Reinsert and replace old BB args.
-  for (BlockArgument ba : block.getArguments().take_front(numOldBBArgs)) {
-    block.addArgument(ba.getType());
-    ba.replaceAllUsesWith(block.getArguments().back());
+  // The only existing argument are for the outputs. Just need to add a new
+  // argument for the outpts and remap the value to use the new argument.
+  for (auto ba : block.getArguments().take_front(numOldBBArgs)) {
+    assert(ba.getType().isa<IREE::Flow::DispatchOutputType>());
+    map.map(ba, block.addArgument(ba.getType()));
+  }
+
+  auto getUsesOfValueOutsideOfDispatchOp =
+      [&](Value v) -> SmallPtrSet<Operation *, 4> {
+    SmallPtrSet<Operation *, 4> res;
+    for (Operation *user : v.getUsers())
+      if (!dispatchOp->isAncestor(user)) res.insert(user);
+    return res;
+  };
+  // Replace all uses of mapped values within the dispatch op to the new value.
+  for (Value value : valuesDefinedAbove) {
+    auto uses = getUsesOfValueOutsideOfDispatchOp(value);
+    value.replaceAllUsesExcept(map.lookup(value), uses);
+  }
+
+  // Clone the marked operations.
+  for (Operation *op : clonedOps) {
+    SmallVector<Value, 2> clonedOpOperands = llvm::to_vector<2>(llvm::map_range(
+        op->getOperands(), [&map](Value v) { return map.lookup(v); }));
+    Operation *insertionPoint = getInsertionPoint(block, clonedOpOperands);
+    if (!insertionPoint) {
+      return op->emitOpError(
+          "failed to clone operation into dispatch workgroup op");
```

Comment:
Be a bit specific can be quite helpful when we hit this. What about something like "unexpected multiple blocks inside dispatch workgroups op"?

Reply:
Thats not the reason for the failure. I actually dont think there should be a reason for failure here. Updated message to say failed to find insertion point.